src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Libero\ContentApiBundle\Adapter;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\LockMode;
+use Doctrine\DBAL\ParameterType;
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\DBAL\Types\Type;
+use IteratorAggregate;
+use Libero\ContentApiBundle\Exception\ItemNotFound;
+use Libero\ContentApiBundle\Exception\UnexpectedVersionNumber;
+use Libero\ContentApiBundle\Exception\VersionNotFound;
+use Libero\ContentApiBundle\Model\ItemId;
+use Libero\ContentApiBundle\Model\ItemListPage;
+use Libero\ContentApiBundle\Model\Items;
+use Libero\ContentApiBundle\Model\ItemVersion;
+use Libero\ContentApiBundle\Model\ItemVersionNumber;
+use Traversable;
+use function array_column;
+use function array_map;
+use function array_pop;
+use function count;
+use function rewind;
+use function sprintf;
+
+final class DoctrineItems implements IteratorAggregate, Items
+{
+    private $connection;
+    private $tableItems;
+    private $tableVersions;
+
+    public function __construct(Connection $connection, string $tablePrefix)
+    {
+        $this->connection = $connection;
+        $this->tableItems = "{$tablePrefix}_items";
+        $this->tableVersions = "{$tablePrefix}_versions";
+    }
+
+    public function add(ItemVersion $item) : void
+    {
+        $this->connection->transactional(
```

Comment:
The default isolation level in MySQL InnoDB is [`REPEATABLE READ`](https://dev.mysql.com/doc/refman/5.7/en/innodb-transaction-isolation-levels.html#isolevel_repeatable-read) which has no surprises for this use case: the transaction sees a snapshot of all data from its beginning. If there are concurrent inserts/updates, besides primary/unique constraints all but one of the transactions should fail on commit and generate a safe 50x error.

Reply:
Same goes for Postgresql, the possible anomaly with the default `REPEATABLE READ` is [phantom reads](https://en.wikipedia.org/wiki/Isolation_(database_systems)#Phantom_reads) but unless we are doing complex `WHERE` statements I wouldn't worry about it.

Reply:
Nope, [`Read Committed is the default isolation level in PostgreSQL`](https://www.postgresql.org/docs/10/transaction-iso.htm). Re-evaluating with this in mind.

Reply:
With `REPEATABLE READ` two concurrent attempts to edit a single `[current_]version` field seem to fail.

Transaction 1:
```
postgres=# BEGIN TRANSACTION;
BEGIN
postgres=# SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
SET
postgres=# SELECT version FROM item WHERE id = 'naledi';                             
 version 
---------
       4
(1 row)

postgres=# UPDATE item SET version=5 WHERE id = 'naledi';
UPDATE 1

```
Before the `COMMIT` start transaction 2:
```
postgres=# BEGIN TRANSACTION;
BEGIN
postgres=# SET TRANSACTION ISOLATION LEVEL REPEATABLE READ;
SET
postgres=# SELECT version FROM item WHERE id = 'naledi';                             
 version 
---------
       4
(1 row)
```

`COMMIT` transaction 1:
```
postgres=# COMMIT;
COMMIT
```

Now try to do an update transaction 2:
```
postgres=# UPDATE item SET version=3 WHERE id = 'naledi';
ERROR:  could not serialize access due to concurrent update
```
In `REPEATABLE READ` transaction 2 can only see rows that were committed before its start, so anything like transaction 1 that commits a concurrent modification to the same field is going to fail transaction 2 and leave the row in a consistent state.

Will try now MySQL which should have `REPEATABLE READ` by default.

Reply:
MySQL needs `SERIALIZABLE` to have the same behavior :man_shrugging:

Reply:
So the transaction isolation levels are inconsistent and probably a dead end here.